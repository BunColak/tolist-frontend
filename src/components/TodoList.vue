<template>
  <div v-if="todoList" class="p-8 w-full">
    <div class="flex items-baseline">
      <h3 class="text-3xl font-bold">{{ todoList.title }}</h3>
      <p class="ml-4 text-sm text-gray-600">
        from
        <router-link class="link" :to="`/templates/${todoList.template.id}`"
          >{{ todoList.template.title }} </router-link
        >by
        <router-link class="link" :to="`/user/${todoList.template.user.id}`">{{
          todoList.template.user.username
        }}</router-link>
      </p>
    </div>
    <div class="mt-8">
      <todo-item
        data-testId="todo"
        v-for="todo in todos"
        :key="todo.id"
        :todo="todo"
        @toggleTodo="toggleTodo"
      />
    </div>
    <div v-if="completedTodos.length" class="mt-8">
      <h4 class="text-xl font-bold">Completed</h4>
      <todo-item
        data-testId="completed-todo"
        v-for="todo in completedTodos"
        :key="todo.id"
        :todo="todo"
        @toggleTodo="toggleTodo"
      />
    </div>
  </div>
</template>

<script>
import TodoListQuery from '../graphql/TodoList.gql'
import ToogleTodoMutation from '../graphql/ToogleTodoMutation.gql'
import TodoItem from './TodoItem.vue'

export default {
  name: 'TodoList',
  components: {
    TodoItem
  },
  apollo: {
    todoList: {
      query: TodoListQuery,
      variables () {
        return {
          id: Number(this.$route.params.id)
        }
      }
    }
  },
  computed: {
    todos () {
      if (this.todoList) {
        return this.todoList.todos
          .filter((t) => !t.completed)
          .sort((a, b) => a.id - b.id)
      }
      return []
    },
    completedTodos () {
      if (this.todoList) {
        return this.todoList.todos
          .filter((t) => t.completed)
          .sort((a, b) => a.id - b.id)
      }
      return []
    },
    listId () {
      return this.$route.params.id
    }
  },
  methods: {
    async toggleTodo (id) {
      try {
        await this.$apollo.mutate({
          mutation: ToogleTodoMutation,
          variables: { id }
        })
        this.$apollo.queries.todoList.refetch()
      } catch (e) {
        console.log(e)
      }
    }
  },
  watch: {
    listId () {
      this.$apollo.queries.todoList.refetch()
    }
  }
}
</script>
