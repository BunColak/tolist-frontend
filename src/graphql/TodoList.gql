query TodoList($id: Float!) {
  todoList(id: $id) {
    id
    title
    todos {
      id
      text
      completed
    }
    template {
      id
      title
      user {
        id
        username
      }
    }
  }
}
