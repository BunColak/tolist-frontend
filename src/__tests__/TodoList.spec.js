import { shallowMount, createLocalVue, RouterLinkStub } from '@vue/test-utils'
import TodoList from '@/components/TodoList'
import VueRouter from 'vue-router'

const localVue = createLocalVue()
const router = new VueRouter()
localVue.use(VueRouter)

const mockTodoList = {
  id: 1,
  title: 'Test Title',
  todos: [
    {
      id: 1,
      text: 'test 1',
      completed: false
    },
    {
      id: 2,
      text: 'test 2',
      completed: false
    },
    {
      id: 3,
      text: 'test 3',
      completed: true
    }
  ],
  template: {
    id: 1,
    title: 'Test Title',
    user: {
      id: 1,
      username: 'TestUsername'
    }
  }
}

describe('TodoList', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(TodoList, {
      propsData: { id: mockTodoList.id },
      data () {
        return {
          todoList: mockTodoList
        }
      },
      localVue,
      router,
      stubs: { RouterLink: RouterLinkStub }
    })
  })

  it('should render without problems', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should group todos properly', () => {
    expect(wrapper.findAll('[data-testId = todo]').length).toBe(2)
    expect(wrapper.findAll('[data-testId = completed-todo]').length).toBe(1)
  })
})
