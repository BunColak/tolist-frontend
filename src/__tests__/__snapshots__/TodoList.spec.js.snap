// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList should render without problems 1`] = `
<div class="p-8 w-full" id="1">
  <div class="flex items-baseline">
    <h3 class="text-3xl font-bold">Test Title</h3>
    <p class="ml-4 text-sm text-gray-600">
      from
      <a class="link">Test Title </a>by
      <a class="link">TestUsername</a>
    </p>
  </div>
  <div class="mt-8">
    <todo-item-stub todo="[object Object]" data-testid="todo"></todo-item-stub>
    <todo-item-stub todo="[object Object]" data-testid="todo"></todo-item-stub>
  </div>
  <div class="mt-8">
    <h4 class="text-xl font-bold">Completed</h4>
    <todo-item-stub todo="[object Object]" data-testid="completed-todo"></todo-item-stub>
  </div>
</div>
`;
